metagraph/src/cli/align.cpp

```diff
@@ -240,6 +241,88 @@ void map_sequences_in_file(const std::string &file,
         delete out;
 }
 
+std::string sequence_to_gfa_path(const std::string &seq,
+                                 const size_t seq_id,
+                                 const std::shared_ptr<DeBruijnGraph> &graph,
+                                 const tsl::ordered_set<uint64_t> &is_unitig_end_node,
+                                 const Config *config) {
+    std::vector<uint64_t> path_nodes;
+    graph->map_to_nodes_sequentially(seq, [&](uint64_t node) {
+        path_nodes.push_back(node);
+    });
+
+    std::string nodes_on_path;
+    std::string cigars_on_path;
+    const size_t overlap = graph->get_k() - 1;
+
+    for (size_t i = 0; i < path_nodes.size() - 1; ++i) {
+        if (config->output_compacted && !is_unitig_end_node.count(path_nodes[i])) {
+            continue;
+        }
+        nodes_on_path += fmt::format("{}+,", path_nodes[i]);
+        cigars_on_path += fmt::format("{}M,", overlap);
+    }
+    uint64_t last_node_to_print = path_nodes.back();
+    // We need to print the id of the last unitig even in the case that
+    // this unitig is not completely covered by the query sequence.
+    while (config->output_compacted && !is_unitig_end_node.count(last_node_to_print)) {
+        uint64_t unique_next_node;
+        graph->adjacent_outgoing_nodes(
+            last_node_to_print,
+            [&](uint64_t node) { unique_next_node = node; }
+        );
+        last_node_to_print = unique_next_node;
+    }
+    nodes_on_path += fmt::format("{}+,", last_node_to_print);
+
+    // Remove right trailing comma.
+    nodes_on_path.pop_back();
+    if (cigars_on_path.size()) {
+        cigars_on_path.pop_back();
+    }
+    return fmt::format("P\t{}\t{}\t{}\n", seq_id, nodes_on_path, cigars_on_path);
+}
+
+void gfa_map_files(const Config *config,
+                   const std::vector<std::string> &files,
+                   const std::shared_ptr<DeBruijnGraph> graph) {
+    logger->trace("Starting GFA mapping:");
+
+    tsl::ordered_set<uint64_t> is_unitig_end_node;
+
+    graph->call_unitigs(
+        [&](const auto &unitig, const auto &path) {
+            (void)unitig;
+            is_unitig_end_node.insert(path.back());
+        }
+    );
+    // Open gfa_file in append mode.
+    std::ofstream gfa_file(utils::remove_suffix(config->gfa_mapping_path, ".gfa") + ".gfa",
+                           std::ios_base::app);
+    std::mutex print_mutex;
+    for (size_t i = 0; i < files.size(); ++i) {
+        const std::string &file = files[i];
```

Comment:
The `i` here is not used anywhere else and is masked by the inner for loop, so this might lead to confusion or bugs later.
```suggestion
    for (const std::string &file : files) {
```